docs/blueprints/EP036.rst

```diff
@@ -0,0 +1,226 @@
+            "table_miss_flow": {
+              "priority": 123, "match": {},
+              "instructions": [{"instruction_type": "goto_table","table_id": 3}]
+            },
+            "napps_table_groups": {"mef_eline": ["evpl"]}
+          },
+          3: {
+            "table_miss_flow": {
+              "priority": 123, "match": {},
+              "instructions": [{"instruction_type": "goto_table","table_id": 5}]
+            },
+            "napps_table_groups": {"mef_eline": ["epl"]}
+          }
+          4: {
+            "table_miss_flow": {},
+            "napps_table_groups": {"telemetry_int": ["base"]}
+          }
+          5: {
+            "table_miss_flow": {},
+            "napps_table_groups": {}
+          }
+        }
+      }
+    }
+
+Events
+======
+
+Subscribed
+-----------
+
+  - **kytos/flow_manager.flow.added**
+  - **kytos/of_core.handshake.completed**
+  - **kytos/flow_manager.flow.error**
+  - **kytos/[mef_eline|telemetry_int|coloring|of_lldp].enable_table** - Every NApp sends an event with their table settings as content.
+
+kytos/<napp_name>.enable_table
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This event is sent from other NApps with information about their table settings. It is empty in case of failure.
+
+.. code:: javascript
+  "content": {
+    "<mef_eline>": {"epl": 3}
+  }
+
+Published
+---------
+
+kytos/multi_table.enable_table
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This event should sent before any NApp sends its first flow to be published.
+
+.. code:: javascript
+
+  "content": {
+    "mef_eline": {"epl": 3, "evpl": 2},
+    "of_lldp": {"base": 0},
+    "coloring": {"base": 0}
+  }
+
+
+
+Endpoint API
+============
+  
+  - GET `/api/kytos/multi_table/v2/` - Get all pipelines in the database. Parameters enabled and disabled are allowed.
+
+  - POST `/api/kytos/multi_table/v2/<pipeline_id>/enable` - Enable pipeline to be applied to all switches
+
+  - POST `/api/kytos/multi_table/v2/<pipeline_id>/disable` - Disable pipeline in case of the user wants to return to the default set up.
+
+  - POST `/api/kytos/multi_table/v2/` - Add pipeline to database. Its body is similar to "pipeline.multi_table".
+
+  - DELETE `/api/kytos/multi_table/v2/<pipeline_id>` - Delete a pipeline from database. It should be disabled.
+
+Enabling pipeline
+==================
+
+Im simple terms, enabling a pipeline will mean to modify every flow so it complies with the pipeline configuration. The process to successfully enable a pipeline is as follows:
+
+  1. **multi_table** publishes `kytos/multi_table.enable_table` event with pipeline content.
+  2. Wait for all NApp to publish their own events with their respective table settings.
+  3. Disable the previous enabled pipeline, if there was any.
+  4. **multi_table** requests all flows from **flow_manager** API.
+  5. **multi_table** duplicates every flow that needs a different table (to achieve this, this NApp needs to be aware of every other NApp cookie to identify ownership).

```

Comment:
Regarding step number 5, since we're at it. Let's also resolve this problem by either:

1 - Introducing an event where each NApp will publish when they're loaded which cookie prefix they're using. 
2 - Augment flow_manager `flows` collection to start supporting also storing `owner: Optional[str]`, `table_group: Optional[str]`.

Option 1 requires fewer adjacent changes and seems not too bad. Wdyt?

Reply:
Yes option 1 seems the best one. Both options imply that the cookies will not change, is this correct?

Reply:
Right. Yes, The cookie won't change, at the moment cookies are always set from the settings when each NApp loads. If one day a network engineer changes that, then either they'll need to delete the old flows or NApps that always send flows like `of_lldp` or `coloring` then it'll get overwritten. When you get flows from `/v2/stored_flows` you can trust the cookie value there.

Reply:
Let's go for it then. Feel free to close the thread once it's in the blueprint. Thanks.

Reply:
Since this event will be introduced it also can solve this other thread here https://github.com/kytos-ng/kytos/pull/359#discussion_r1160879295, Aldo.

Reply:
Introduced `kytos/<napp_name>.cookie` event under Subscribed. I am going to referenced it in the steps.

Reply:
Sounds good. On this event content are we going to introduce which table_groups this NApp supports? I was also looking for that since we discussed that during pipeline creation all table_groups will be known upfront by `of_multi_table`. Let me know if that changed or if you're thinking about other way to cover this.  

```
"content": {
    "of_lldp": "0xAB"
  }
````

Reply:
But in this cookie initial event shouldn't it also include which table groups that it supports? Otherwise, when a pipeline is first created how will `of_multi_table` identify an invalid  table_group on `POST /api/kytos/of_multi_table/v1/pipeline`? I'm under the impression that `of_multi_table` would need to know the cookie upfront (that it'll need later on to identify table changes), but also which table groups eventually will be able to be set since we discussed that we wouldn't have an error when `of_multi_table` publishes `kytos/of_multi_table.enable_table`, right? Let me know if I misunderstood something.

@Alopalao, I just realized option 1 also has a limitation that if just one table group changes, like `mef_eline` "evpl" to a different table when both are initially in the same table, since the cookie prefix is per NApp and not on table group, `of_multi_table` won't be able to differentiate which flows belong to `evpl` or `epl`. Option 1 could still work but it would end up reinstalling and deleting more flows than what it should, so it wouldn't be optimal. We might need to reconsider option 2. If we go for option too then we don't even need to know about cookies since you'd be able to group by `table_group`. We still have time to consider which option, let me know if you're also seeing this drawback and if we'll need to go for option 2 augmenting `flow_manager` to also support `owner: Optional[str], table_group: Optional[str]` on flow entries.

Reply:
what do you mean by invalid table_group? I haven't thought of any pipeline that would cause an error other that not following the linked list rules, number table going up and keywords like `base`, `EVPL`, `vlan_id` and more. Any table that follows those rules is able to be enable?
`evpl` and `epl` can be recognize by the content of `"flow.match"`.

Reply:
Got it. We have a limited amount of possible group tables like "epl" or "ip_proto". We can have them specified in the `openapi` file. But I think it is worth to explain them in the blueprint as well. If we are going to add a `table_group` field, a flow can match more than just one table group. Does this means that we will need to have priority for each one? A flow can have the label "evpl" and "tcp_src" right?

Reply:
Nevermind the priority we could just match first table number with the first found table group that matches, but I think table_group should be a set though.

Reply:
> We can have them specified in the openapi file. But I think it is worth to explain them in the blueprint as well

For sure. That's a good idea to validate them at request level with string enums. Let's go for it. 

>  If we are going to add a table_group field, a flow can match more than just one table group. 

Nope. Essentially, from my understanding, each flow should belong to only a single table_group. They're just a logical exclusive set of flows, which represents a category of high level match functionality (which is a set of flows that network engineers will have the freedom to move between tables based on their pipeline). 

> Does this means that we will need to have priority for each one? A flow can have the label "evpl" and "tcp_src" right?

That's a good point. Nope. The flow priority each NApp will set based on how they are already doing. `"evpl"` will a table group, `"tcp_src"` is a match field. If one day "tcp_src' needs to be part of "evpl" then it's an implementation match detail on `mef_eline` that shouldn't impact `of_multi_table`. But if one day "tcp_src" needs to be treated as a table_group then a new table group would be introduced which should give meaning to a new category. Ideally, on `of_multi_table` we probably want to only deal with higher level table groups just so you don't have to deal with match implementation details.

Reply:
> but I think table_group should be a set though.

Check out my previous reply. I believe we're discussing about the same thing but you were also seeing low level match fields as a table group, which they can turned into (if they are sufficiently large to represent a category of relevant matches) but by default they won't. 

Let me know if that makes sense or if you're seeing that differently or seeing any other issues.

Reply:
I was thinking more on the process to modify (if needed) the table of a flow. We could have a thread for each switch although I think is going to take me time so it could be a future spec.

Reply:
> I was thinking more on the process to modify (if needed) the table number of a flow. We could have a thread for each switch although I think is going to take me time so it could be a future spec.

Ultimately, on Kytos core, each message to a socket on `msg_out` queue will be handled asynchronously, and each connection handles each socket. On this NApp the only way you can control and take advantage of IO concurrency is how you send events or request to `flow_manager`. It might be beneficial to group by switches in the batches sent to `flow_manager` (like mef_eline does) but it's not recommended to add new application managed threads at the NApp level, pretty much the core and all NApps are evolving towards `asyncio`. 

> I think is going to take me time so it could be a future spec.

The easiest lowest effort (based on what's currently available) approach would be to leverage `kytos.flow_manager.install|delete` like mef_eline does, otherwise without any concurrency sending request synchronously for every like 50 or 100 flows with 500+ flows can add up quickly to the overall process. This is a critical path, non-functional requirement, for this NApp. Sending it all synchronously to flow_manager considering the number of flows this NApp might be dealing wouldn't be optimal (although `flow_manager` will dispatch the FlowMods in an async queue anyway). I'll recommend that you consider at least this event based approach if you think you'd be short on time based on what you have for this version. Let me know if you can also cover this part.

Reply:
Understood. Also, I think the main topic of this conversation was resolved.